EmailResponder/FeedbackDecryptor/redactor.py

```diff
@@ -0,0 +1,356 @@
+
+    obj = _generate_ios_vpn_feedback(171, log)
+    redact_sensitive_values(obj)
+
+    # The order of fields in the JSON string may change due to reserialization # during redaction.
+    expectedOutputOrdering1 = _generate_ios_vpn_feedback(171, 'ExtensionInfo: {"PacketTunnelProvider": {"Event": "Start", "StartMethod": "Container"}}')
+    expectedOutputOrdering2 = _generate_ios_vpn_feedback(171, 'ExtensionInfo: {"PacketTunnelProvider": {"StartMethod": "Container", "Event": "Start"}}')
+
+    assert(obj == expectedOutputOrdering1 or
+           obj == expectedOutputOrdering2)
+    
+    # Test where no redaction attempts are made based on the client version
+
+    obj = _generate_ios_vpn_feedback(1, log)
+    obj_copy = _generate_ios_vpn_feedback(1, log)
+    redact_sensitive_values(obj)
+    assert(obj == obj_copy)
+
+    # Test where no redaction attempts are made based on the client platform
+
+    obj = _generate_android_feedback(171, log)
+    obj_copy = _generate_android_feedback(171, log)
+    redact_sensitive_values(obj)
+    assert(obj == obj_copy)
+
+    print 'redact_sensitive_values_ios_vpn_test okay'
+
+
+def redactors(client_platform, client_version):
+    '''
+    Return redactors to use for the target client platform and version.
+    '''
+    redactors = [_redact_upstream_proxy_errors]
+    if client_platform == "ios-vpn" and client_version >= 160:
+        redactors.append(_ios_vpn_redact_start_tunnel_with_options)
+    return redactors
+
+
+def run_redactors(obj, redactors):
+    '''
+    Traverse the object and perform any necessary redactions.
+    '''
+    for path, val in utils.objwalk(obj):
+        for redactor in redactors:
+            redactor(obj, path, val)
+
+diagnostic_msg_regex = re.compile(r'([a-zA-Z]+): ({.*})')
+
+
+def _redact_upstream_proxy_errors(obj, path, val):
+    '''
+    Redacts any text which follows the target upstream proxy error string.
+    '''
+    if isinstance(val, utils.string_types):
+
+        target = "upstreamproxy error: proxyURI url.Parse: parse "
+
+        index = val.find(target)
+        if index == -1:
+            return
+
+        result = diagnostic_msg_regex.match(val)
+        if result is not None:
+            try:
+                j = json.loads(result.group(2))
+                found = _redact_target_from_dict(target, j)
+                if found is True:
+                    redacted_val = result.group(1) + ": " + json.dumps(j)
+                    utils.assign_value_to_obj_at_path(obj, path, redacted_val)
+                    return
+            except ValueError:
+                pass
+       
+        # Fallback on a less finessed redaction
+        redacted_val = val[:index+len(target)] + "<redacted>"
+        utils.assign_value_to_obj_at_path(obj, path, redacted_val)
+
+
+def _redact_target_from_dict(target, d):
+    '''
+    Redact the target string from any values in the dictionary.
```

Comment:
I think the docstring should explain that the function redacts everything _after_ target in a string. I was confused.